docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,294 @@
+ ┌─────────────────┐
+ │ArgoCD API Server│
+ └────────┬────────┘
+          │
+          │ GET http://extension-host.com:8080
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+
+```
+
+- **Example 2**:
+
+If a backend provides an API under the `/apiv1/metrics` endpoint, ArgoCD
+should be able to invoke it such as:
+
+```
+   ┌────────────┐
+   │ ArgoCD UI  │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension/apiv1/metrics/123
+          │
+          ▼
+ ┌─────────────────┐
+ │ArgoCD API Server│
+ └────────┬────────┘
+          │
+          │ GET http://extension-host.com:8080/apiv1/metrics/123
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+##### Considerations
+
+- The `idleConnTimeout` can be used to avoid accumulating too many
+  goroutines waiting slow for extensions. In this case a proper timeout
+  error (408) should be returned to the browser.
+- Headers, scheme, http verb and request body are forwarded as it is
+  received by the API server to the backend service.
+- A new header is added in the forwared request (`X-Forwarded-Host`) to
+  allow ssl redirection
+
+----
+
+#### [UC-2]: As an ArgoCD admin, I want to define extensions rbacs so access permissions can be enforced
+
+Extend ArgoCD rbac registering a new `ResourceType` for extensions in the
+[policy configuration][3]. The current policy permission configuration is
+defined as:
+
+```
+p, <subject>, <resource>, <action>, <object>, <access>
+```
+
+With a new resource type for extensions, admins will be able to configure
+access rights per extension per project. The `object` field must contain
+the project name and the extension name in the format
+`<project>/<extension>`
+
+- **Example 1**:
+
+```
+p, role:allow-extensions, extensions, *, some-project/some-extension, allow
+```
+
+In the example 1, a permission is configured to allowing the subject
+`role:allow-extensions`, for the resource type `extensions`, for all (`*`)
+actions, in the project `some-project`, for the extension name
+`some-extension`.
+
+
+- **Example 2**:
+
+```
+p, role:allow-extensions, extensions, *, */some-extension, allow
+```
+
+In the example 2, the permission is similar to the example 1 with the
+difference that the extension `some-extension` will be allowed for all
+projects.
+
+- **Example 3**:
+
+```
+p, role:allow-extensions, extensions, *, */*, allow
+```
+
+In the example 3, the subject `role:allow-extensions` is allowed to
+execute extensions in all projects.
+
+### Security Considerations
+
+- ArgoCD API Server must apply **authn** and **authz** for all incoming
+  extensions requests
+- ArgoCD must authorize requests coming from UI and check that the
+  authenticated user has access to invoke a specific URL belonging to an
+  extension
+
+### Risks and Mitigations
+
+### Upgrade / Downgrade
+
+## Drawbacks
+
+- Slight increase in ArgoCD code base complexity.
```

Comment:
* increased security risk
* impact of extensions on other uses (mitigated by rate limiting + idle conn timeout)